be/dtm0_log.h

```diff
@@ -0,0 +1,233 @@
+ *    m0_be_tx_open(tx, cred);
+ *
+ *    ...
+ *    m0_be_dtm0_log_update(log, tx, txr);
+ *    tx_close(tx);
+ * }
+ *
+ * 3. When transaction become persistent on particular participant, the state
+ *    of the transaction for this participant will be updated as
+ *    M0_DTML_STATE_PERSISTENT.
+ *
+ * {
+ *    struct m0_be_tx_credit     cred;
+ *    struct m0_be_dtm0_log     *log;
+ *    struct m0_be_tx           *tx;
+ *    struct m0_dtm0_log_record *txr;
+ *    struct m0_be_seg          *seg;
+ *
+ *    m0_be_dtm0_log_credit(op, tx, seg, &cred);
+ *    m0_be_tx_open(tx, cred);
+ *
+ *    ...
+ *    m0_be_dtm0_log_update(log, tx, txr);
+ *    tx_close(tx);
+ * }
+ *
+ * 4. When transaction become persistent on remote participant, the participant
+ *    sends the persistent notice to rest of the participants and originator
+ *    stating that transaction is persistent on its store. Upon receiving
+ *    persistent notice each of the participants and originator will update the
+ *    state of transaction for remote participant as M0_DTML_STATE_PERSISTENT
+ *    in DTM0 log record.
+ *
+ *    There is one special case where a process got a persistent notice from
+ *    another process but the corresponding request and reply were not
+ *    received/processed yet. In this case, the DTM service should add a special
+ *    log record: a log record with empty payload (NULL). The users of DTM0 log
+ *    should be ready to encounter such a log entry, and treat it differently
+ *    whenever it is required.
+ *
+ * {
+ *    struct m0_be_tx_credit     cred;
+ *    struct m0_be_dtm0_log     *log;
+ *    struct m0_be_tx           *tx;
+ *    struct m0_dtm0_log_record *txr;
+ *    struct m0_be_seg          *seg;
+ *
+ *    m0_be_dtm0_log_credit(op, tx, seg, &cred);
+ *    m0_be_tx_prep(tx, &cred);
+ *    m0_be_tx_open(tx);
+ *
+ *    ...
+ *    m0_be_dtm0_log_update(log, tx, txr);
+ *    tx_close(tx);
+ * }
+ *
+ *
+ */
+
+/* Participant state */
+enum m0_dtm0_log_pa_state {
+	M0_DTML_STATE_INPROGRESS,
+	M0_DTML_STATE_EXECUTED,
+	M0_DTML_STATE_PERSISTENT,
+};
+
+enum m0_be_dtm0_log_credit_op {
+	M0_DTML_CREATE,
+	M0_DTML_SENT,
+	M0_DTML_EXECUTED,
+	M0_DTML_PERSISTENT,
+	M0_DTML_REDO
+};
+
+/* Unique identifier for request */
+struct m0_dtm0_dtx_id {
+	struct m0_fid di_fid;
+	uint64_t      di_ts;
+};
+
+struct m0_dtm0_log_pa {
+	struct m0_fid             pfid;
+	enum m0_dtm0_log_pa_state pstate;
+};
+
+/* TODO: define in a separate header. */
+struct m0_dtm0_txr {
+	struct m0_dtm0_dtx_id  dt_tid;
+	struct m0_dtm0_log_pa *dt_participants;
+	uint16_t               dt_participants_nr;
+	struct m0_buf          dt_txr_payload;
+};
+
+struct m0_dtm0_log_record {
+	uint64_t               dlr_magic;
+	struct m0_dtm0_txr     dlr_txr;
+	struct m0_be_list_link dlr_link; /* link into m0_be_dtm0_log::list */
+	struct m0_list_link    dlr_tlink;
+};
+
+struct m0_be_dtm0_log {
+	struct m0_mutex    dl_lock;  /* volatile structure */
```

Comment:
Probably you'd need to save a pointer to the be segment here. Still it can be done outside DLD work.

Reply:
Will do outside DLD whenever required, will keep in mind.